zilliqa/src/transaction.rs

```diff
@@ -334,6 +362,94 @@ impl SignedTransaction {
             }
         }
     }
+
+    pub fn validate(
+        &self,
+        account: &Account,
+        block_gas_limit: EvmGas,
+        chain_id: u64,
+    ) -> Result<bool> {
+        self.validate_input_size()?;
+        self.validate_gas_limit(block_gas_limit)?;
+        self.validate_chain_id(chain_id)?;
+        self.validate_sender_account(account)
+    }
+
+    fn validate_input_size(&self) -> Result<bool> {
+        let (input_size, tx_kind) = match self {
+            SignedTransaction::Legacy { tx, .. } => (tx.input.len(), tx.to),
+            SignedTransaction::Eip2930 { tx, .. } => (tx.input.len(), tx.to),
+            SignedTransaction::Eip1559 { tx, .. } => (tx.input.len(), tx.to),
+            _ => return Ok(true),
+        };
+
+        if input_size > EVM_MAX_TX_INPUT_SIZE {
+            warn!(
+                "Evm transaction input size: {input_size} exceeds limit: {EVM_MAX_TX_INPUT_SIZE}"
+            );
+            return Ok(false);
+        }
+
+        if tx_kind == TxKind::Create && input_size > EVM_MAX_INIT_CODE_SIZE {
+            warn!("Evm transaction initcode size: {input_size} exceeds limit: {EVM_MAX_INIT_CODE_SIZE}");
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+
+    fn validate_gas_limit(&self, block_gas_limit: EvmGas) -> Result<bool> {
+        if self.gas_limit() > block_gas_limit {
+            warn!("Transaction gas limit exceeds block gas limit!");
+            return Ok(false);
+        }
+
+        if let SignedTransaction::Zilliqa { tx, .. } = self {
+            if tx.gas_limit < ZIL_MIN_GAS_UNITS {
+                warn!("Insufficient gas give for zil transaction, given: {}, required: {ZIL_MIN_GAS_UNITS}!", tx.gas_limit);
+                return Ok(false);
+            }
+        }
+
+        let gas_limit = self.gas_limit();
+
+        if gas_limit < EVM_MIN_GAS_UNITS {
+            warn!("Insufficient gas give for evm transaction, given: {gas_limit}, required: {EVM_MIN_GAS_UNITS}!");
+            return Ok(false);
```

Comment:
Can we return a reason code here? It would be massively useful for application developers to know what has caused their txn to fail to be accepted.